### ISSUE TITLE
Incorrect result when using `showCategory` option in `dotplot` function

### ISSUE THREAD

**Issue by randef1ned** [YuLab-SMU/enrichplot#219]
Hello Professor Yu,

When using the `dotplot` function in the `enrichplot` package, I noticed that the results displayed in the plot are not consistent with my own implementation of the function. Further investigation revealed that this is due to a problem with the way the `showCategory` option is being implemented. When this option is changed, the results in the plot are altered, which is clearly a major issue.

Steps to reproduce:
```library(org.Rn.eg.db)
library(tidyverse)
library(clusterProfiler)
library(enrichplot)
gobp <- gseGO(sorted_genes, ont = 'BT', OrgDb = org.Rn.eg.db, pvalueCutoff = 0.4) %>%
  simplify(cutoff = 0.7, by = 'p.adjust', select_fun = min)
plot_gobp <- dotplot(showCategory = 15, x = 'GeneRatio', title = 'GO (BP)', label_format = 100, order = TRUE)
```

When I changed the `showCategory` to 25, the first enriched pathway is not the same as the plot of 15's. 

<img width="899" alt="showCategory = 15" src="https://user-images.githubusercontent.com/46381867/208432343-32257eb0-69db-41cb-9c4c-cc8bee631624.png">
<img width="900" alt="showCategory = 25" src="https://user-images.githubusercontent.com/46381867/208432365-1d3803fe-cd1d-4121-a733-9d77612bb039.png">

I am using the latest version of the enrichplot package (version 1.18.3) and R version 4.2.2 on Fedora 37. I am happy to provide any additional information or assistance that may be necessary to help resolve this issue.

**Comment by Baboon61** [comment]
I would also like to have a comment from the devs on this `showCategory` behavior. Thanks

**Comment by erzakiev** [comment]
Yes it seems like some terms are skipped when using `showCategory`

**Comment by zqun1** [comment]
Yes, I had the same problem. Dotplot is not showing my top (highest geneRatio) processes and skips some processes. Took me a while to figure out the problem until I found this posted issue. Any solution to this?

**Comment by randef1ned** [comment]
Unfortunately I cannot reproduce this issue on Windows. So, is this a Linux-specific problem?

I can reproduce this error on MacOS. Below are code snippets:

```
data(geneList, package='DOSE')
de <- names(geneList)[1:100]
yy <- enrichKEGG(de, pvalueCutoff=1)
dotplot(yy, showCategory = 5, x = 'GeneRatio')
```
<img width="1052" height="510" alt="Image" src="https://github.com/user-attachments/assets/dbbec249-dfa5-4a8b-9558-b1df003f929f" />

```
dotplot(yy, showCategory = 10, x = 'GeneRatio')
```
<img width="1052" height="1016" alt="Image" src="https://github.com/user-attachments/assets/89823131-1afe-4160-a6ca-ae11efea7426" />

Actually I cannot identify the problem. Do you guys have any ideas?

**Comment by zqun1** [comment]
> Unfortunately I cannot reproduce this issue on Windows. So, is this a Linux-specific problem?
> 
> I can reproduce this error on MacOS. Below are code snippets:
> 
> ```
> data(geneList, package='DOSE')
> de <- names(geneList)[1:100]
> yy <- enrichKEGG(de, pvalueCutoff=1)
> dotplot(yy, showCategory = 5, x = 'GeneRatio')
> ```
> 
> <img alt="Image" width="1052" height="510" src="https://private-user-images.githubusercontent.com/46381867/495175923-dbbec249-dfa5-4a8b-9558-b1df003f929f.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTkxNTE5MzUsIm5iZiI6MTc1OTE1MTYzNSwicGF0aCI6Ii80NjM4MTg2Ny80OTUxNzU5MjMtZGJiZWMyNDktZGZhNS00YThiLTk1NTgtYjFkZjAwM2Y5MjlmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTI5VDEzMTM1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVkMDUwZGFjOTZkY2FkYzM3OGZkYjc4YjIwNTI5Zjk5ODMxMWYyYjk4YjQ3N2ZmMDVlNjMwNmFkYzM4Yjk4MGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.O6UDvd0trtPI3Jbvt95xr2nePKy8psVXE3fAYfXx8hI">
> 
> ```
> dotplot(yy, showCategory = 10, x = 'GeneRatio')
> ```
> 
> <img alt="Image" width="1052" height="1016" src="https://private-user-images.githubusercontent.com/46381867/495175742-89823131-1afe-4160-a6ca-ae11efea7426.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTkxNTE5MzUsIm5iZiI6MTc1OTE1MTYzNSwicGF0aCI6Ii80NjM4MTg2Ny80OTUxNzU3NDItODk4MjMxMzEtMWFmZS00MTYwLWE2Y2EtYWUxMWVmZWE3NDI2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTI5VDEzMTM1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEyYzQ4ZDM3ZDI4NDQyNWU2ZjhiYzJhZWU5YzZmNzk5YzEwNWY5OTRhMWJmY2UyYWNlMGI4YmNkM2UyZDc4MjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.WJsbAYRqMecict9ZTHEfn77tGMAqygDAXPICtyXs1K0">
> 
> Actually I cannot identify the problem. Do you guys have any ideas?

I had this problem exactly in Windows R.

**Comment by edna-stewart** [comment]
I also have this problem in Linux (R studio hosted on AWS/Ubuntu environment)

**Comment by GuangchuangYu** [comment]
Fixed on the current devel branch. This issue had the same root cause as the more recent `orderBy`/`showCategory` reports: `dotplot()` used to let `fortify()` truncate the rows for numeric `showCategory` before the final ordering was applied, so increasing `showCategory` could change the leading categories in ways that did not match the requested ranking variable. The selection path now orders the full fortified data first and only then applies numeric `showCategory`, and regression coverage verifies that the leading IDs stay stable when `showCategory` increases. Closing this as fixed.
